File: 259-3sum-smaller/259-3sum-smaller.js
/**
 * @param {number[]} nums
 * @param {number} target
 * @return {number}
 */
var threeSumSmaller = function(nums, target) {
    
    nums.sort((a,b) => a - b)
    
    function twoSumSmaller(startIdx, target){
        let count = 0;
        let left = startIdx
        let right = nums.length - 1
        while(left < right){
            if(nums[left] + nums[right] < target){
                count += right - left
                left++;
            } else {
                right--;
            }
        }
        return count
    }
    
    let res = 0;
    for(let i = 0; i < nums.length; i++){
        res += twoSumSmaller(i+1, target - nums[i])
    }
    
    return res
    
};

File: 259-3sum-smaller/259-3sum-smaller.py
class Solution:
    def threeSumSmaller(self, nums: List[int], target: int) -> int:
        
        nums.sort()
        
        def twoSumSmaller(startIdx, target):
            twoSumCount = 0
            left = startIdx
            right = len(nums) - 1
            while left < right:
                if nums[right] + nums[left] < target:
                    twoSumCount += right - left
                    left += 1
                else:
                    right -= 1
            return twoSumCount
        
        threeSumCount = 0
        for i in range(len(nums) - 2):
            threeSumCount += twoSumSmaller(i + 1, target - nums[i])
         
        return threeSumCount
    
# Time: O(n^2)
# Space: O(1)